core/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -219,6 +232,9 @@ public TopDocs call() throws Exception {
                     topDocsCollector = TopScoreDocCollector.create(numDocs, lastEmittedDoc);
                 }
                 collector = topDocsCollector;
+                if (doProfile) {
+                    collector = new InternalProfileCollector(collector, CollectorResult.REASON_SEARCH_TOP_HITS, Collections.emptyList());
```

Comment:
to prevent all these if statements I wonder if we should add a `CollectorBuilder` class that is used instead of `Collector collector` and has methods like
`CollectorBuilder#set(Collector collector)` `CollectorBuilder#add(List<Collector> collectors)` and `Collector wrap(Collector c)` that is simply called all the time for each new collector so we have only one place where we check if we profile and we never forget to add one?